pkg/cloud/rhos/rhos.go

```diff
@@ -0,0 +1,204 @@
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/stolostron/submariner-addon/pkg/cloud/reporter"
+
+	"github.com/submariner-io/cloud-prepare/pkg/api"
+	cloudpreparerhos "github.com/submariner-io/cloud-prepare/pkg/rhos"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	cloudsYAMLName = "clouds.yaml"
+	cloudName      = "cloud"
+	gwInstanceType = "PnTAE.CPU_4_Memory_8192_Disk_50"
+)
+
+type rhosProvider struct {
+	infraID              string
+	nattPort             uint16
+	routePort            string
+	gatewayMetricsPort   uint16
+	globalnetMetricsPort uint16
+	cloudPrepare         api.Cloud
+	reporter             api.Reporter
+	gwDeployer           api.GatewayDeployer
+	gateways             int
+	nattDiscoveryPort    int64
+}
+
+func NewRHOSProvider(
+	restMapper meta.RESTMapper,
+	kubeClient kubernetes.Interface,
+	dynamicClient dynamic.Interface,
+	hubKubeClient kubernetes.Interface,
+	eventRecorder events.Recorder,
+	region, infraID, clusterName, credentialsSecretName, instanceType string,
+	nattPort, nattDiscoveryPort, gateways int) (*rhosProvider, error) {
+	if infraID == "" {
+		return nil, fmt.Errorf("cluster infraID is empty")
+	}
+
+	if nattPort == 0 {
+		nattPort = constants.SubmarinerNatTPort
+	}
+
+	if instanceType != "" {
+		instanceType = gwInstanceType
+	}
+
+	if gateways < 1 {
+		return nil, fmt.Errorf("the count of gateways is less than 1")
+	}
+
+	if nattDiscoveryPort == 0 {
+		nattDiscoveryPort = constants.SubmarinerNatTDiscoveryPort
+	}
+
+	projectID, cloudEntry, providerClient, err := newClient(hubKubeClient, clusterName, credentialsSecretName)
+	if err != nil {
+		klog.Errorf("Unable to retrieve the rhosclient :%v", err)
+		return nil, err
+	}
+	k8sClient := k8s.NewInterface(kubeClient)
+
+	cloudInfo := cloudpreparerhos.CloudInfo{
+		InfraID:   infraID,
+		Region:    region,
+		K8sClient: k8sClient,
+		Client:    providerClient,
+	}
+
+	cloudPrepare := cloudpreparerhos.NewCloud(cloudInfo)
+	msDeployer := ocp.NewK8sMachinesetDeployer(restMapper, dynamicClient)
+
+	gwDeployer := cloudpreparerhos.NewOcpGatewayDeployer(cloudInfo, msDeployer, projectID, gwInstanceType,
+		"", cloudEntry, true)
+
+	return &rhosProvider{
+		infraID:              infraID,
+		nattPort:             uint16(nattPort),
+		routePort:            strconv.Itoa(constants.SubmarinerRoutePort),
+		gatewayMetricsPort:   constants.SubmarinerGatewayMetricsPort,
+		globalnetMetricsPort: constants.SubmarinerGlobalnetMetricsPort,
+		cloudPrepare:         cloudPrepare,
+		gwDeployer:           gwDeployer,
+		reporter:             reporter.NewEventRecorderWrapper("RHOSCloudProvider", eventRecorder),
+		nattDiscoveryPort:    int64(nattDiscoveryPort),
+		gateways:             gateways,
+	}, nil
+}
+
+// PrepareSubmarinerClusterEnv prepares submariner cluster environment on RHOS
+// The below tasks will be executed
+// 1. create the inbound and outbound firewall rules for submariner, below ports will be opened
+//    - IPsec IKE port (by default 500/UDP)
```

Comment:
```suggestion

```
Please remove this line as we are not opening the 500/UDP Port. In its place, you can add ESP, AH ports